CONTRIBUTING.md

```diff
@@ -0,0 +1,121 @@
+When reporting an issue, please include a precise reproduction in the bug report
+when that's possible, as it is a very useful tool to investigate. You should
+also check that you are using the latest version of OCaml-lsp and that a similar
+issue has not already been submitted.
+
+## Code contributions
+
+### Styleguides
+
+- **OCaml**: a large part of the code base is written in OCaml and the project
+  is configured to work with
+  [ocamlformat](https://ocaml.org/p/ocamlformat/latest) (version defined in the
+  [`.ocamlformat` file](.ocamlformat)).
+
+- **TypeScript**: TypeScript is used to describe certain end-to-end tests
+  (abbreviated as `e2e`) and the project uses the
+  [prettier](https://prettier.io/) formatter.
+
+Apart from that, the project tries to apply implicit conventions, at the
+decision of the maintainers. At the same time, it tries to follow certain naming
+conventions:
+
+- use of `t` manifest types in modules, when it makes sense;
+- conversion functions respecting the naming scheme: `t_of_xxx` or `xxx_of_t`.
+  For example, when converting to or from data described by
+  [Yojson](https://ocaml.org/p/yojson/latest), use `t_of_yojson` and
+  `yojson_of_t`.
+
+Changes unrelated to the issue addressed by a PR should be made in a separate
+PR. Additionally, formatting changes in parts of the code not concerned by a
+specific PR should be proposed in another PR.
+
+Ideally, any opened issue should be accompanied by a test with a reproduction.
+When working on a fix for an issue, the first commit should contain the test
+showing the issue. Following commits should fix the issue and update the test
+result accordingly.
+
+### Repository organization
+
+The repository exposes a number of separate libraries (some of which are
+internal) and vendor libraries (to reduce the dependencies required by the
+project). Here is a list of the libraries exposed by the project.
+
+#### ocaml-lsp-server
+
+Contains the concrete implementation of a protocol server language for OCaml. A
+frontend used in particular by [Visual Studio
+Code](https://github.com/ocamllabs/vscode-ocaml-platform), but also by code
+editors supporting LSP. The code lives mainly in the following directories:
+[ocaml-lsp-server/](ocaml-lsp-server/).
+
+In addition, the project exposes two sub-directories dedicated to [code
+actions](ocaml-lsp-server/src/code_actions) and [custom
+requests](https://github.com/ocaml/ocaml-lsp/tree/master/ocaml-lsp-server/src/custom_requests).
+
+In most cases, it is likely that the contributions will focus solely on this
+project.
+
+##### Warning
+
+For historical reasons, but also for development convenience, `ocaml-lsp-server`
+should not build logic based on `Parsetree` (which changes from version to
+version and migration logic is not provided by `lsp`, nor `ocaml-lsp-server` but
+by [Merlin](https://github.com/ocaml/merlin)). If a command, or a constant,
+relies on the `Parsetree`, it is **strongly** recommended to build a command in
```

Comment:
I think you meant the typed tree rather than the parse tree. The parse tree is already essentially a public interface as it's used by hundreds of ppx's. Ocamllsp isn't any special here.

Also, I would weaken the language a little. It's certainly fine to experiment with things, and the downstream projects that have asked to maintain this discipline don't always follow this rule themselves. I'd say it's just recommended to be done with care and marked as unstable. In particular, it should be fairly easy to get rid of the feature if it proves intrusive or hard to maintain.

Reply:
You are right ! Thinking about "dealing with parse tree, avoiding typed tree", i mixed the two.